File: c/face.c
#include <opencv2/opencv.hpp>
#include <stdio.h>

using namespace cv;

// Dummy function for LED brightness control - to be replaced with actual control logic
void set_leds_brightness(int brightness) {
    // Placeholder: Implement LED control logic here
    printf("LED brightness set to: %d\n", brightness);
}

int main() {
    VideoCapture cap(0); // Open the default camera
    if (!cap.isOpened()) {
        printf("Error: Could not open camera.\n");
        return -1;
    } else {
        printf("camera detected\n");
    }

    CascadeClassifier faceCascade;
    if (!faceCascade.load("/usr/share/opencv4/haarcascades/haarcascade_frontalface_default.xml")) {
        printf("Error: Could not load face cascade.\n");
        return -1;
    } else {
        printf("cascade module loaded \n");
    }

    Mat frame, gray;
    std::vector<Rect> faces;
    double knownDistance = 30.48; // Known distance from camera to face in cm (1 foot)
    double knownWidth = 14.0;     // Known width of the face in cm
    double focalLength = 0.0;     // Focal length (initially unknown)

    while (true) {
        cap >> frame; // Capture a frame
        if (frame.empty()) break; // Check for failure

        cvtColor(frame, gray, COLOR_BGR2GRAY); // Convert to grayscale
        faceCascade.detectMultiScale(gray, faces, 1.1, 4, 0, Size(30, 30)); // Detect faces

        if (faces.size() == 0) {
            set_leds_brightness(0); // Turn off LEDs if no faces are detected
        }

        for (size_t i = 0; i < faces.size(); i++) {
            Rect face = faces[i];
            rectangle(frame, Point(face.x, face.y), Point(face.x+face.width, face.y+face.height), Scalar(255, 0, 0), 2);

            // Calculate focal length (once)
            if (focalLength == 0.0) {
                focalLength = (face.width * knownDistance) / knownWidth;
            }

            // Calculate distance to face
            double distance = (knownWidth * focalLength) / face.width;
            printf("Distance to face: %.2f cm\n", distance);

            // Dummy call for LED brightness control
            int brightness = static_cast<int>((1 - (distance / knownDistance)) * 255);
            set_leds_brightness(brightness);
        }

        if (waitKey(30) >= 0) break; // Exit loop if any key is pressed
    }

    cap.release(); // When everything done, release the video capture object
    return 0;
}

File: c/face.py
import cv2
import numpy as np
import board
import neopixel

# Configuration for NeoPixel
LED_PIN = board.D18
NUM_LEDS = 79  # Adjust based on your setup
BRIGHTNESS = 1.0
ORDER = neopixel.GRB

# Initialize NeoPixel
pixels = neopixel.NeoPixel(LED_PIN, NUM_LEDS, brightness=BRIGHTNESS, auto_write=False, pixel_order=ORDER)

def calculate_brightness(distance, max_distance):
    """Calculate brightness based on distance, inversely proportional."""
    return max(0, min(1, 1 - (distance / max_distance)))

def set_leds_brightness(brightness):
    """Set the brightness of all LEDs."""
    color = (int(255 * brightness), int(255 * brightness), int(255 * brightness))  # White color, adjust brightness
    for i in range(NUM_LEDS):
        pixels[i] = color
    pixels.show()

# Load the cascade for face detection
face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + '/usr/share/opencv4/haarcascades/haarcascade_frontalface_default.xml')

# To capture video from the webcam
cap = cv2.VideoCapture(0)

# Known parameters (adjust these based on your setup)
known_distance = 30.48  # known distance from the camera in cm
known_width = 14.0     # known width of the face in cm
focal_length = 0.0     # will be calculated

try:
    while True:
        # Read the frame
        _, img = cap.read()
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

        # Detect the faces
        faces = face_cascade.detectMultiScale(gray, 1.1, 4)
        
        if len(faces) == 0:
            # If no faces detected, turn off LEDs
            set_leds_brightness(0)
        else:
            for (x, y, w, h) in faces:
                # Calculate focal length (once)
                if focal_length == 0.0:
                    focal_length = (w * known_distance) / known_width

                # Calculate distance to face
                distance = (known_width * focal_length) / w
                print(f"Distance to face: {distance:.2f} cm")

                # Adjust LED brightness based on distance
                brightness = calculate_brightness(distance, known_distance)
                set_leds_brightness(brightness)

        # Stop if escape key is pressed (in terminal, consider a different stopping mechanism)
        if cv2.waitKey(30) == 27:
            break
finally:
    # Release the VideoCapture object and turn off LEDs
    cap.release()
    set_leds_brightness(0)  # Ensure LEDs are turned off when stopping
